### PR TITLE
Splitted ParserDelegate for Request and Response interfaces and updated unit tests

### PR DESCRIPTION
**Pull request by sergey-litvinov**
Hi Benjamin,
Here is a code based on your feedback for #11 issue. 
Could you please review it?

### CODE REVIEW

File: src/HttpMachine/rl/HttpParser.cs.rl [hunk 1]
```diff
@@ -163,12 +170,13 @@ namespace HttpMachine
 
         action leave_query_string {
             //Console.WriteLine("leave_query_string fpc " + fpc + " qsMark " + qsMark);
```
Comment (bvanderveen) [inline src/HttpMachine/rl/HttpParser.cs.rl:172]: Should this `EnsureRequestParser()`?

File: tests/HttpMachine.Tests/HttpMachineTests.cs [hunk 2]
```diff
@@ -475,8 +529,18 @@ static void ThreeChunkScan(IEnumerable<TestRequest> requests)
...
+                        Handler handler;
+                        HttpParser parser;
+                        if (isRequest)
+                        {
+                            handler = new RequestHandler();
+                            parser = new HttpParser((RequestHandler)handler);
+                        }
+                        else
+                        {
+                            handler = new ResponseHandler();;
```
Comment (bvanderveen) [inline tests/HttpMachine.Tests/HttpMachineTests.cs]: Two semicolons!

File: README.md [hunk 3]
```diff
@@ -20,7 +20,21 @@ MIT License. See LICENSE.txt.
 
 ## Usage
 
-HttpMachine provides HTTP data through callbacks. To receive these callbacks, implement the `IHttpParserHandler` interface.
+HttpMachine provides HTTP data through callbacks. To receive these callbacks, implement on of the `IHttpRequestParserDelegate` or `IHttpResponseParserDelegate` interfaces.
```
Comment (bvanderveen) [inline README.md]: Small copy edit—make it read:

"To receive these callbacks, implement the `IHttpRequestParserDelegate` or `IHttpResponseParserDelegate` interfaces."

(remove "on of")

**Comment by bvanderveen** [comment]
This looks great! I made just a few notes. Fix those up and I will merge! Thank you Sergey!

File: README.md (hunk 3 above)
Comment (bvanderveen) [inline README.md]: Oh I see what you meant: "To receive these callbacks, implement either the `IHttpRequestParserDelegate` or the `IHttpResponseParserDelegate` interface."

File: README.md [hunk 4]
```diff
@@ -20,7 +20,21 @@ MIT License. See LICENSE.txt.
...
+        void OnFragment(HttpParser parser, string fragment);
+        void OnQueryString(HttpParser parser, string queryString);
+    }
+
+    public interface IHttpResponseParserDelegate : IHttpParserDelegate
+    {
+        void OnResponseCode(HttpParser parser, int statusCode, string statusReason); 
+    }
 
     public interface IHttpParserHandler
```
Comment (bvanderveen) [inline README.md:39]: This bit in the readme should be renamed "IHttpParserDelegate" and updated to match the new definition in the source code.

File: README.md [hunk 5]
```diff
@@ -39,7 +53,7 @@ HttpMachine provides HTTP data through callbacks. To receive these callbacks, im
 
 Then, create an instance of `HttpParser`. Whenever you read data, execute the parser on the data. The `Execute` method returns the number of bytes successfully parsed. If value is not the same as the length of the buffer you provided, an error occurred while parsing. Make sure you provide a zero-length buffer at the end of the stream, as some callbacks may still be pending.
 
-    var handler = new MyHttpParserHandler();
+    var handler = new MyHttpRequestParserHandler();
```
Comment (bvanderveen) [inline README.md]: Maybe `MyHttpParserDelegate`?

File: README.md (hunk 3 above)
Comment (sergey-litvinov) [inline README.md]: Updated readme to the text from last comment. My english isn't so good :)

File: README.md (hunk 5 above)
Comment (sergey-litvinov) [inline README.md]: updated.

File: README.md (hunk 4 above)
Comment (sergey-litvinov) [inline README.md:39]: What do you mean? I specified here all three actual interfaces - base one, and two inheritors

File: src/HttpMachine/rl/HttpParser.cs.rl (hunk 1 above)
Comment (sergey-litvinov) [inline src/HttpMachine/rl/HttpParser.cs.rl:172]: I think it should be `EnsureResponseParser`. Action `on_response_message` will be called by parser when there will be response code and response message, so they are related to response, and require Response handler

File: tests/HttpMachine.Tests/HttpMachineTests.cs (hunk 2 above)
Comment (sergey-litvinov) [inline tests/HttpMachine.Tests/HttpMachineTests.cs]: Oops, sorry :) Fixed

File: README.md (hunk 4 above)
Comment (bvanderveen) [inline README.md:39]: No worries, I'll make the change. Thanks!

**Comment by sergey-litvinov** [comment]
Thank you!

**Pull request merged**